docs/decisions/0045-function-call-behavior.md

```diff
@@ -0,0 +1,322 @@
+            AvailableFunctions = null,
+            RequiredFunctions = null,
+        };
+    }
+}
+```
+
+To satisfy the "connector/mode-agnostic model" driver, the behavior should not be configured on model-specific prompt execution setting classes, such as `OpenAIPromptExecutionSettings`, as it is done today. Instead, it should be configured on a model-agnostic one - `PromptExecutionSettings`.
+
+```csharp
+PromptExecutionSettings settings = new() { FunctionChoiceBehavior = FunctionChoiceBehavior.RequiredFunctionChoice() };
+```
+
+### Sequence diagram
+<img src="./diagrams/tool-behavior-usage-by-ai-service.png" alt="Tool choice behavior usage by AI service.png" width="600"/>
+
+### [Out of scope]Breaking glass support
+The list of choice classes described above may not be sufficient to support all scenarios users might encounter. To accommodate these, the `FunctionCallChoice.Configure` method accepts an instance of model a connector use internally, allowing users to access and modify it from within the config method of a custom function call choice.
+```csharp
+// Custom function call choice
+public sealed class NewCustomFunctionChoiceBehavior : FunctionChoiceBehavior
+{
+    public override FunctionChoiceBehaviorConfiguration GetConfiguration(FunctionChoiceBehaviorContext context)
+    {
+        var model = context.Model;
+
+        // The CompletionsOptions, ChatCompletionsToolChoice, etc are data model classes used by the OpenAIChatCompletionService connector internally.
+        ((CompletionsOptions)model).ToolChoice = new ChatCompletionsToolChoice(new FunctionDefinition("NEW-TOOL-CHOICE-MODE"));
+        ((CompletionsOptions)model).Tools.Add(new ChatCompletionsFunctionToolDefinition(<functions-to-advertise>);
+ 
+        return new FunctionChoiceBehaviorConfiguration()
+        {
+            Model = model; // Return the model back to the calling connector to indicate that we control the function call choice ourselves, and there is no need to apply the mapping logic connector side that would be applied otherwise.
+            MaximumAutoInvokeAttempts = this.MaximumAutoInvokeAttempts,
+            MaximumUseAttempts = this.MaximumUseAttempts,
+            AllowAnyRequestedKernelFunction = false
+        };
+    }
+}
+...
+
+// Registering the custom choice
+PromptExecutionSettings settings = new() { FunctionChoiceBehavior = new NewCustomFunctionChoiceBehavior() };
+```
+
+### Support in JSON prompts
+Considering the hierarchical nature of the choice behavior model classes, polymorphic deserialization should be enabled for instances when functional choice behavior needs to be configured in JSON prompts - MD and Kernel (config.json) prompts, as well as YAML prompts - handlebars.
+```json
+{
+    ...
+    "execution_settings": {
+        "default": {
+            "temperature": 0.4,
+            "function_choice_behavior": {
+                "type": "auto", //possible values - auto, required, none
+                "functions": [
```

Comment:
Is the following scenario correct? In the event that someone defines the functions in the JSON/yaml config and specifies the type as "auto?" In this example "plugin1-function1" is the only tool made available to the model even though the kernel has plugin1-function1, plugin2-function1, plugin3-function1?

If someone first configures the functions in the config, and later updates the filter to include other plugins/functions and these two fully qualified function lists now diverge, what functions list has higher precedence?

Reply:
> Is the following scenario correct? In the event that someone defines the functions in the JSON/yaml config and specifies the type as "auto?" In this example "plugin1-function1" is the only tool made available to the model even though the kernel has plugin1-function1, plugin2-function1, plugin3-function1?

Yes, that's the intention: whoever configures it explicitly states that only the 'plugin1-function1' function should be advertised and that the `auto` function choice behavior should be used, even though the other functions are available on kernel.

> If someone first configures the functions in the config, and later updates the filter to include other plugins/functions and these two fully qualified function lists now diverge, what functions list has higher precedence?

I would say that whatever is configured in the filter should have higher precedence over the configuration.